 # Instructions:
 - For all cells marked with <font color="red">(CODE Needed)</font> below, replace `#CODE_HERE` with your solution


# 1. Load Spark Context (No CODE Needed)

In [1]:
from pyspark import SparkContext, SparkConf
myconf = SparkConf() \
        .setMaster("local[*]") 
        #.set("spark.driver.cores",4) \
        #.set("spark.executor.memory","2g") 
sc = SparkContext(conf=myconf)

# 2. Load some printing functions (No CODE Needed)

In [2]:
#These are some printing helper functions we will use to make the output more clear
from pprint import pprint
def title(s):
    pprint("---- %s -----" %s)    
    
def see(s, v):
    pprint("---- %s -----" %s)
    pprint(v)

# 3. Load each file in an RDD and add the year to the data (No CODE Needed)

In [24]:
def createOneYearRdd(year):
    file = "../data/flight-data/csv/%s"%year+"*"
    see("year:"+str(year), file)
    fileRDD =sc.textFile(file)
    linesRDD = fileRDD.filter(lambda x:  not x.startswith("DEST_")) \
              .filter(lambda line: line.count(',')==2)
    arrRDD = linesRDD.map(lambda line: (line.split(","), year))
    see("rdd count for year:"+str(year), arrRDD.count())
    title("rdd elements for year:"+str(year))
    for arr in (arrRDD.take(5)):
        print(arr)
    return arrRDD

separateRDDs  = list(map(createOneYearRdd, range(2010,2016)))
                 
see("arryay of RDDs", separateRDDs)

'---- year:2010 -----'
'../data/flight-data/csv/2010*'
'---- rdd count for year:2010 -----'
253
'---- rdd elements for year:2010 -----'
(['United States', 'Romania', '1'], 2010)
(['United States', 'Ireland', '264'], 2010)
(['United States', 'India', '69'], 2010)
(['Egypt', 'United States', '24'], 2010)
(['Equatorial Guinea', 'United States', '1'], 2010)
'---- year:2011 -----'
'../data/flight-data/csv/2011*'
'---- rdd count for year:2011 -----'
253
'---- rdd elements for year:2011 -----'
(['United States', 'Saint Martin', '2'], 2011)
(['United States', 'Guinea', '2'], 2011)
(['United States', 'Croatia', '1'], 2011)
(['United States', 'Romania', '3'], 2011)
(['United States', 'Ireland', '268'], 2011)
'---- year:2012 -----'
'../data/flight-data/csv/2012*'
'---- rdd count for year:2012 -----'
243
'---- rdd elements for year:2012 -----'
(['United States', 'Croatia', '1'], 2012)
(['United States', 'Ireland', '252'], 2012)
(['Egypt', 'United States', '13'], 2012)
(['United States', 'India', '

# 4. Merge all RDDs into one Rdd  (No CODE Needed)

In [4]:
allYearsRdd  =sc.union(separateRDDs)
see("allYearsRDD count", allYearsRdd.count())
see("allYearsRDD sample", allYearsRdd.takeSample(False, 10, 17))

'---- allYearsRDD count -----'
1488
'---- allYearsRDD sample -----'
[(['United States', 'Saint Barthelemy', '39'], 2011),
 (['Egypt', 'United States', '13'], 2011),
 (['Egypt', 'United States', '13'], 2013),
 (['United States', 'Poland', '33'], 2015),
 (['Haiti', 'United States', '201'], 2014),
 (['United States', 'Canada', '8177'], 2014),
 (['United States', 'Netherlands', '607'], 2012),
 (['United States', 'Saint Lucia', '121'], 2010),
 (['Italy', 'United States', '381'], 2012),
 (['Greece', 'United States', '17'], 2012)]


# 5. Compute an RDD containing all flights which departed from  the US  <font color="red">(CODE Needed)</font>
- Hint: make 1 `filter` transformation

In [5]:
fromUSRdd = allYearsRdd.filter(lambda v: v[0][1]=='United States')

see("fromUSRdd count",fromUSRdd.count())

see("fromUSRdd take(10)",fromUSRdd.take(10))

'---- fromUSRdd count -----'
765
'---- fromUSRdd take(10) -----'
[(['Egypt', 'United States', '24'], 2010),
 (['Equatorial Guinea', 'United States', '1'], 2010),
 (['Costa Rica', 'United States', '477'], 2010),
 (['Senegal', 'United States', '29'], 2010),
 (['Guyana', 'United States', '17'], 2010),
 (['Malta', 'United States', '1'], 2010),
 (['Bolivia', 'United States', '46'], 2010),
 (['Anguilla', 'United States', '21'], 2010),
 (['Turks and Caicos Islands', 'United States', '136'], 2010),
 (['Saint Vincent and the Grenadines', 'United States', '1'], 2010)]


### Expectated output for step 5.

<pre>
'---- fromUSRdd count -----'
765
'---- fromUSRdd take(10) -----'
[(['Egypt', 'United States', '24'], 2010),
 (['Equatorial Guinea', 'United States', '1'], 2010),
 (['Costa Rica', 'United States', '477'], 2010),
 (['Senegal', 'United States', '29'], 2010),
 (['Guyana', 'United States', '17'], 2010),
 (['Malta', 'United States', '1'], 2010),
 (['Bolivia', 'United States', '46'], 2010),
 (['Anguilla', 'United States', '21'], 2010),
 (['Turks and Caicos Islands', 'United States', '136'], 2010),
 (['Saint Vincent and the Grenadines', 'United States', '1'], 2010)]
</pre>

# 6. Create an RDD with a composite key (destination, year) and a value n which is the corresponding number of flights  as an integer <font color="red">(CODE Needed)</font>
- Use the `int()` function to convert a string to an integer
- Hint: make 1 `map` transformation

In [6]:
compositeKeyRdd =  fromUSRdd.map(lambda v: ((v[0][0],v[1]),int(v[0][2])))

see("compositeKeyRdd count",compositeKeyRdd.count())

see("compositeKeyRdd take(10)",compositeKeyRdd.take(10))

'---- compositeKeyRdd count -----'
765
'---- compositeKeyRdd take(10) -----'
[(('Egypt', 2010), 24),
 (('Equatorial Guinea', 2010), 1),
 (('Costa Rica', 2010), 477),
 (('Senegal', 2010), 29),
 (('Guyana', 2010), 17),
 (('Malta', 2010), 1),
 (('Bolivia', 2010), 46),
 (('Anguilla', 2010), 21),
 (('Turks and Caicos Islands', 2010), 136),
 (('Saint Vincent and the Grenadines', 2010), 1)]


### Exepected output for step 6
<pre>
'---- compositeKeyRdd count -----'
765
'---- compositeKeyRdd take(10) -----'
[(('Egypt', 2010), 24),
 (('Equatorial Guinea', 2010), 1),
 (('Costa Rica', 2010), 477),
 (('Senegal', 2010), 29),
 (('Guyana', 2010), 17),
 (('Malta', 2010), 1),
 (('Bolivia', 2010), 46),
 (('Anguilla', 2010), 21),
 (('Turks and Caicos Islands', 2010), 136),
 (('Saint Vincent and the Grenadines', 2010), 1)]
 </pre>

# 7. Sum all the flights for the same destination and the same year <font color="red">(CODE Needed)</font>
- Hint: make 1 `reduceByKey` transformation

In [7]:
groupedRdd = compositeKeyRdd.reduceByKey(lambda v1,v2: v1+v2)

see("groupedRdd count",groupedRdd.count())

see("groupedRdd take(10)",groupedRdd.take(10))

'---- groupedRdd count -----'
765
'---- groupedRdd take(10) -----'
[(('Ethiopia', 2010), 12),
 (('Fiji', 2010), 53),
 (('Federated States of Micronesia', 2010), 46),
 (('French Guiana', 2010), 4),
 (('Netherlands', 2010), 586),
 (('Denmark', 2010), 98),
 (('Belgium', 2010), 408),
 (('Guatemala', 2010), 386),
 (('Vietnam', 2010), 1),
 (('South Korea', 2010), 683)]


### Exepected output for step  7 
<pre>
'---- groupedRdd count -----'
765
'---- groupedRdd take(10) -----'
[(('Ethiopia', 2010), 12),
 (('Fiji', 2010), 53),
 (('Federated States of Micronesia', 2010), 46),
 (('French Guiana', 2010), 4),
 (('Netherlands', 2010), 586),
 (('Denmark', 2010), 98),
 (('Belgium', 2010), 408),
 (('Guatemala', 2010), 386),
 (('Vietnam', 2010), 1),
 (('South Korea', 2010), 683)]
 </pre>

# 8. Sort the summed data by the highest sum <font color="red">(CODE Needed)</font>
- Hint: make 1 `map` transformation, 1 `sortByKey(ascending=False)`, then 1 `map`


In [8]:
sortedByCountRdd = groupedRdd.map(lambda v: (v[1],v[0])).sortByKey(ascending=False).map(lambda v: (v[1],v[0]))

see("sortedByCountRdd count",sortedByCountRdd.count())

see("sortedByCountRdd take(10)",sortedByCountRdd.take(10))


'---- sortedByCountRdd count -----'
765
'---- sortedByCountRdd take(10) -----'
[(('United States', 2015), 370002),
 (('United States', 2014), 358354),
 (('United States', 2011), 352742),
 (('United States', 2010), 348113),
 (('United States', 2012), 347452),
 (('United States', 2013), 343132),
 (('Canada', 2011), 8514),
 (('Canada', 2015), 8399),
 (('Canada', 2010), 8271),
 (('Canada', 2012), 8034)]


### Exepected output for step  8
<pre>
'---- sortedByCountRdd count -----'
765
'---- sortedByCountRdd take(10) -----'
[(('United States', 2015), 370002),
 (('United States', 2014), 358354),
 (('United States', 2011), 352742),
 (('United States', 2010), 348113),
 (('United States', 2012), 347452),
 (('United States', 2013), 343132),
 (('Canada', 2011), 8514),
 (('Canada', 2015), 8399),
 (('Canada', 2010), 8271),
 (('Canada', 2012), 8034)]
 </pre>

# 9. Compute the Sum of all Trips <font color="red">(CODE Needed)</font>
- Hint: make 1 `map` , 1 `sum`


In [9]:
sumAllTrips = sortedByCountRdd.map(lambda v: (v[1])).sum()
see("sumAllTrips", sumAllTrips)

'---- sumAllTrips -----'
2352144


### Exepected output for step  9
<pre>
'---- sumAllTrips -----'
2352144
 </pre>

# 10. Compute the Sum of Each Year's Trip <font color="red">(CODE Needed)</font>
- Hint: make 1 `map` , 1 `reduceByKey`, 1 `collectAsMap`


In [10]:
TPY = sortedByCountRdd.map(lambda v: (v[0][1],v[1])).reduceByKey(lambda v1,v2: v1+v2).collectAsMap()

see("TPY", TPY)



'---- TPY -----'
{2010: 385434,
 2011: 390613,
 2012: 385244,
 2013: 380985,
 2014: 397960,
 2015: 411908}


### Exepected output for step  10
<pre>
'---- TPY -----'
{2010: 385434,
 2011: 390613,
 2012: 385244,
 2013: 380985,
 2014: 397960,
 2015: 411908}
 </pre>

# 11. Use the above sums to compute percent Per Year, percent from Total for each country, year pair <font color="red">(CODE Needed)</font>
- Hint: make 1 `map`
- use the function `percent`
- use TPY as a lookup table


In [11]:
def percent(number, total):
    return 100.0* number/float(total)

stats = sortedByCountRdd.map(lambda v: (v[0],v[1], percent(v[1],TPY[v[0][1]]), percent(v[1],sum(TPY.values()))))

see("stats count",stats.count())

see("stats take(10)",stats.take(10))

'---- stats count -----'
765
'---- stats take(10) -----'
[(('United States', 2015), 370002, 89.8263689950183, 15.730414464420546),
 (('United States', 2014), 358354, 90.04774349180822, 15.235206688025903),
 (('United States', 2011), 352742, 90.30472616118767, 14.99661585345115),
 (('United States', 2010), 348113, 90.31714898011073, 14.799816677890469),
 (('United States', 2012), 347452, 90.19011327885704, 14.771714656925766),
 (('United States', 2013), 343132, 90.06443823247635, 14.588052432164018),
 (('Canada', 2011), 8514, 2.1796509588774566, 0.36196763463461423),
 (('Canada', 2015), 8399, 2.039047554308244, 0.3570784781884102),
 (('Canada', 2010), 8271, 2.145892682015598, 0.3516366344917658),
 (('Canada', 2012), 8034, 2.0854315706409445, 0.3415607207721976)]


### Exepected output for step  11
<pre>
'---- stats count -----'
765
'---- stats take(10) -----'
[(('United States', 2015), 370002, 89.8263689950183, 15.730414464420546),
 (('United States', 2014), 358354, 90.04774349180822, 15.235206688025903),
 (('United States', 2011), 352742, 90.30472616118767, 14.99661585345115),
 (('United States', 2010), 348113, 90.31714898011073, 14.799816677890469),
 (('United States', 2012), 347452, 90.19011327885704, 14.771714656925766),
 (('United States', 2013), 343132, 90.06443823247635, 14.588052432164018),
 (('Canada', 2011), 8514, 2.1796509588774566, 0.36196763463461423),
 (('Canada', 2015), 8399, 2.039047554308244, 0.3570784781884102),
 (('Canada', 2010), 8271, 2.145892682015598, 0.3516366344917658),
 (('Canada', 2012), 8034, 2.0854315706409445, 0.3415607207721976)]
 </pre>

# 12. Transform the above table into the format below <font color="red">(CODE Needed)</font>
- Hint: make 1 `map`, 1 `groupByKey`, 1 `mapValues` 
- you will also need the python functions `list`, `sorted`, `dict`
- use TPY as a lookup table


In [12]:
preTable = stats.map(lambda v: (v[0][0],(v[0][1],(v[1],v[2],v[3])))).groupByKey().sortByKey()\
.mapValues(lambda v: dict(list(v)))


see("preTable count",preTable.count())

see("preTable take(5)",preTable.take(5))

'---- preTable count -----'
165
'---- preTable take(5) -----'
[('Afghanistan',
  {2010: (11, 0.002853925704530477, 0.0004676584426803801),
   2011: (8, 0.0020480629164927946, 0.0003401152310402764),
   2012: (5, 0.0012978787469759426, 0.00021257201940017278)}),
 ('Algeria',
  {2013: (2, 0.0005249550507237817, 8.50288077600691e-05),
   2014: (9, 0.002261533822494723, 0.000382629634920311),
   2015: (4, 0.0009710906318886742, 0.0001700576155201382)}),
 ('Angola',
  {2010: (14, 0.0036322690784933347, 0.0005952016543204838),
   2011: (13, 0.0033281022393007913, 0.0005526872504404493),
   2012: (12, 0.003114908992742262, 0.0005101728465604147),
   2013: (12, 0.0031497303043426907, 0.0005101728465604147),
   2014: (13, 0.003266659965825711, 0.0005526872504404493),
   2015: (15, 0.003641589869582528, 0.0006377160582005184)}),
 ('Anguilla',
  {2010: (21, 0.0054484036177400025, 0.0008928024814807257),
   2011: (21, 0.005376165155793586, 0.0008928024814807257),
   2012: (19, 0.004931939238508582

### Exepected output for step  12
<pre>
'---- preTable count -----'
165
'---- preTable take(10) -----'
[('Afghanistan',
  {2010: (11, 0.002853925704530477, 0.0004676584426803801),
   2011: (8, 0.0020480629164927946, 0.0003401152310402764),
   2012: (5, 0.0012978787469759426, 0.00021257201940017278)}),
 ('Algeria',
  {2013: (2, 0.0005249550507237817, 8.50288077600691e-05),
   2014: (9, 0.002261533822494723, 0.000382629634920311),
   2015: (4, 0.0009710906318886742, 0.0001700576155201382)}),
 ('Angola',
  {2010: (14, 0.0036322690784933347, 0.0005952016543204838),
   2011: (13, 0.0033281022393007913, 0.0005526872504404493),
   2012: (12, 0.003114908992742262, 0.0005101728465604147),
   2013: (12, 0.0031497303043426907, 0.0005101728465604147),
   2014: (13, 0.003266659965825711, 0.0005526872504404493),
   2015: (15, 0.003641589869582528, 0.0006377160582005184)}),
 ('Anguilla',
  {2010: (21, 0.0054484036177400025, 0.0008928024814807257),
   2011: (21, 0.005376165155793586, 0.0008928024814807257),
   2012: (19, 0.004931939238508582, 0.0008077736737206565),
   2013: (22, 0.0057745055579616, 0.0009353168853607602),
   2014: (34, 0.008543572218313398, 0.0014454897319211748),
   2015: (41, 0.00995367897685891, 0.0017430905590814169)}),
 ('Antigua and Barbuda',
  {2010: (123, 0.03191207833247715, 0.00522927167724425),
   2011: (146, 0.0373771482259935, 0.006207102966485045),
   2012: (145, 0.03763848366230233, 0.006164588562605011),
   2013: (123, 0.03228473561951258, 0.00522927167724425),
   2014: (115, 0.028897376620765906, 0.0048891564462039735),
   2015: (126, 0.030589354904493236, 0.005356814888884354)})]
 </pre>

# 13. Transform the above table into the format below<font color="red">(CODE Needed)</font>
- Hint: make 1 `map`, you will need more than one line in the map, so create a mapping function

In [42]:
def loopDic(cntry,dct):
#    tempL = keysList
#    tempJ = keysList
#    tempK = keysList
#    i = 0
#    while i < len(keysList):
#        tempL[i]=('PcntTotal-' + str(keysList[i]) , pcntList[i][2])
#        tempJ[i]=('PcntYear-' + str(keysList[i]) , pcntList[i][1])
#        tempK[i]=('Tot-' + str(keysList[i]) , pcntList[i][0])
#        i+=1

    out = {"Country":cntry}
    for y,vs in dct.items():
        out['PcntTotal-' + str(y)] = vs[2]
        out['PcntYear-' + str(y)] = vs[1]
        out['Tot-' + str(y)] = vs[0]        
#    tempL = tempL.extend(tempJ)
#    tempL = tempL.extend(tempK)
#     cntry = ('Country',cntry)    
#     tempL.insert(0,cntry)

    return out

Table = preTable.map(lambda v:  loopDic(v[0], v[1]))

see("Table count",Table.count())

see("Table take(3)",Table.take(3))




'---- Table count -----'
165
'---- Table take(3) -----'
[{'Country': 'Afghanistan',
  'PcntTotal-2010': 0.0004676584426803801,
  'PcntTotal-2011': 0.0003401152310402764,
  'PcntTotal-2012': 0.00021257201940017278,
  'PcntYear-2010': 0.002853925704530477,
  'PcntYear-2011': 0.0020480629164927946,
  'PcntYear-2012': 0.0012978787469759426,
  'Tot-2010': 11,
  'Tot-2011': 8,
  'Tot-2012': 5},
 {'Country': 'Algeria',
  'PcntTotal-2013': 8.50288077600691e-05,
  'PcntTotal-2014': 0.000382629634920311,
  'PcntTotal-2015': 0.0001700576155201382,
  'PcntYear-2013': 0.0005249550507237817,
  'PcntYear-2014': 0.002261533822494723,
  'PcntYear-2015': 0.0009710906318886742,
  'Tot-2013': 2,
  'Tot-2014': 9,
  'Tot-2015': 4},
 {'Country': 'Angola',
  'PcntTotal-2010': 0.0005952016543204838,
  'PcntTotal-2011': 0.0005526872504404493,
  'PcntTotal-2012': 0.0005101728465604147,
  'PcntTotal-2013': 0.0005101728465604147,
  'PcntTotal-2014': 0.0005526872504404493,
  'PcntTotal-2015': 0.0006377160582005184,

### Exepected output for step  13
<pre>
'---- Table count -----'
165
'---- Table take(3) -----'
[{'Country': 'Afghanistan',
  'PcntTotal-2010': 0.0004676584426803801,
  'PcntTotal-2011': 0.0003401152310402764,
  'PcntTotal-2012': 0.00021257201940017278,
  'PcntYear-2010': 0.002853925704530477,
  'PcntYear-2011': 0.0020480629164927946,
  'PcntYear-2012': 0.0012978787469759426,
  'Tot-2010': 11,
  'Tot-2011': 8,
  'Tot-2012': 5},
 {'Country': 'Algeria',
  'PcntTotal-2013': 8.50288077600691e-05,
  'PcntTotal-2014': 0.000382629634920311,
  'PcntTotal-2015': 0.0001700576155201382,
  'PcntYear-2013': 0.0005249550507237817,
  'PcntYear-2014': 0.002261533822494723,
  'PcntYear-2015': 0.0009710906318886742,
  'Tot-2013': 2,
  'Tot-2014': 9,
  'Tot-2015': 4},
 {'Country': 'Angola',
  'PcntTotal-2010': 0.0005952016543204838,
  'PcntTotal-2011': 0.0005526872504404493,
  'PcntTotal-2012': 0.0005101728465604147,
  'PcntTotal-2013': 0.0005101728465604147,
  'PcntTotal-2014': 0.0005526872504404493,
  'PcntTotal-2015': 0.0006377160582005184,
  'PcntYear-2010': 0.0036322690784933347,
  'PcntYear-2011': 0.0033281022393007913,
  'PcntYear-2012': 0.003114908992742262,
  'PcntYear-2013': 0.0031497303043426907,
  'PcntYear-2014': 0.003266659965825711,
  'PcntYear-2015': 0.003641589869582528,
  'Tot-2010': 14,
  'Tot-2011': 13,
  'Tot-2012': 12,
  'Tot-2013': 12,
  'Tot-2014': 13,
  'Tot-2015': 15}]
 </pre>

# 14. Optional: Convert the data to a Pandas Dataframe <font color="green">(No CODE Needed)</font>


In [43]:
import pandas as pd

pd.DataFrame(data=Table.collect())

Country  PcntTotal-2010  PcntTotal-2011  PcntTotal-2012  \
0                 Afghanistan        0.000468        0.000340        0.000213   
1                     Algeria             NaN             NaN             NaN   
2                      Angola        0.000595        0.000553        0.000510   
3                    Anguilla        0.000893        0.000893        0.000808   
4         Antigua and Barbuda        0.005229        0.006207        0.006165   
5                   Argentina        0.007823        0.007780        0.008843   
6                       Aruba        0.015263        0.014752        0.014838   
7                   Australia        0.012329        0.011904        0.011776   
8                     Austria        0.001531        0.001445        0.001445   
9                  Azerbaijan        0.000043             NaN             NaN   
10                    Bahrain        0.001275        0.001701        0.002763   
11                   Barbados        0.005527        0.005484        0.005144   
12                    Belarus             NaN        0.000043        0.000043   
13                    Belgium        0.017346        0.015985        0.014965   
14                     Belize        0.005144        0.004804        0.004804   
15                    Bermuda        0.007780        0.008120        0.008078   
16                    Bolivia        0.001956        0.002593        0.001488   
17                     Brazil        0.042302        0.041196        0.041622   
18     British Virgin Islands        0.002083        0.003104        0.003529   
19                     Brunei             NaN             NaN             NaN   
20                   Bulgaria        0.000043        0.000170        0.000255   
21               Burkina Faso             NaN             NaN             NaN   
22                    Burundi             NaN             NaN             NaN   
23                   Cameroon             NaN             NaN        0.000043   
24                     Canada        0.351637        0.361968        0.341561   
25                 Cape Verde        0.000765        0.000553        0.000808   
26             Cayman Islands        0.010501        0.010671        0.010969   
27                       Chad             NaN             NaN             NaN   
28                      Chile        0.007057        0.007568        0.007568   
29                      China        0.019046        0.021427        0.021640   
..                        ...             ...             ...             ...   
135              Sint Maarten        0.002593        0.010203        0.010331   
136                  Slovakia        0.000043             NaN             NaN   
137           Solomon Islands             NaN             NaN        0.000085   
138              South Africa        0.001063        0.001020        0.001020   
139               South Korea        0.029037        0.030313        0.034819   
140                     Spain        0.017941        0.018919        0.016581   
141                  Suriname        0.000510        0.000468        0.000595   
142                    Sweden        0.002763        0.002508        0.002126   
143               Switzerland        0.013392        0.013350        0.012117   
144                    Taiwan        0.011691        0.010544        0.009736   
145                  Tanzania             NaN             NaN             NaN   
146                  Thailand        0.000680        0.000510        0.000255   
147               The Bahamas        0.038391        0.034479        0.041452   
148                The Gambia             NaN        0.000043        0.000085   
149                      Togo             NaN        0.000043             NaN   
150       Trinidad and Tobago        0.007950        0.008715        0.009183   
151                   Tunisia             NaN             NaN        0.000085   
152                    Turkey        0.003189        